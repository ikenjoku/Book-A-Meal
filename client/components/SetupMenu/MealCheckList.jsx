import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';


export const MealCheckList = ({ meals, handleSelect, isFetching }) => {
  return (isFetching ? 
    <h3>Fetching meals...</h3>
    :
    meals.map(meal => (
    <div key={meal.id}>
      <label>
        <input
          type="checkbox"
          name="meals"
          value={meal.id}
          onClick={handleSelect}
          />
          <span>{meal.name}</span>
      </label>
    </div>
  )));
};

export const mapStateToProps = state => ({
  isFetching: state.setupMenuReducer.isFetching,
});

MealCheckList.propTypes = {
  meals: PropTypes.arrayOf(PropTypes.object).isRequired,
  handleSelect: PropTypes.func.isRequired,
  isFetching: PropTypes.bool.isRequired,
};

export default connect(mapStateToProps, null)(MealCheckList);
