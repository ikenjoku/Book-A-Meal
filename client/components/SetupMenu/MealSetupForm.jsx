import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Pagination from 'react-js-pagination';
import { connect } from 'react-redux';
import { getAMeals } from '../../actions/setupMenuActions';
import MealCheckList from './MealCheckList';

export class MealSetupForm extends Component {
  state = {
    date: this.props.date || '',
    mealIds: [],
    activePage: 1,
  };

  handleSubmit = (event) => {
    event.preventDefault();
    this.props.handleSubmit({
      date: this.state.date,
      mealIds: this.state.mealIds,
    });
  }

  handleSelect = (event) => {
    if (event.target.checked) {
      this.setState({
        mealIds: [
          ...this.state.mealIds,
          event.target.value,
        ],
      });
    } else {
      const currentMealIds = this.state.mealIds.filter(mealId =>
        mealId !== event.target.value);

      this.setState({ mealIds: currentMealIds });
    }
  }

  handlePageChange = (pageNumber) => {
    const limit = 5;
    this.props.getAMeals({ limit, page: pageNumber });
    this.setState(() => ({ activePage: pageNumber }));
  }

  render() {
    return (
      <form
        className="menu-setup-form"
        onSubmit={this.handleSubmit}
      >
        <div>
          <button className="calender-btn">
            <i className="far fa-calendar-alt fa-2x" />
          </button>
          <input
            readOnly
            name="date"
            type="date"
            value={this.state.date}
          />
        </div>
        <h3>Choose meals to add to menu</h3>
        <div className="meal-check-list">
          <MealCheckList
            meals={this.props.meals}
            handleSelect={this.handleSelect}
          />
        </div>
        <div className="center">
          <Pagination
            activePage={this.state.activePage}
            itemsCountPerPage={5}
            totalItemsCount={this.props.count}
            pageRangeDisplayed={5}
            onChange={this.handlePageChange}
          />
        </div>
        <div className="setup-btn-container">
          <button
            value="Submit"
            type="submit"
          >
            {this.props.action} Menu
          </button>
        </div>
      </form>
    );
  }
}

MealSetupForm.propTypes = {
  date: PropTypes.string.isRequired,
  getAMeals: PropTypes.func.isRequired,
  count: PropTypes.number.isRequired,
};

export const mapStateToProps = state => ({
  count: state.setupMenuReducer.count,
  pages: state.setupMenuReducer.pages,
});

MealSetupForm.propTypes = {
  meals: PropTypes.arrayOf(PropTypes.object).isRequired,
  handleSubmit: PropTypes.func.isRequired,
  action: PropTypes.string.isRequired,
};

export default connect(mapStateToProps, { getAMeals })(MealSetupForm);
