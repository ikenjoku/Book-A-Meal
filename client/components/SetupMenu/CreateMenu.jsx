import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import MealSetupForm from './MealSetupForm';
import { getAMeals, setAMenu } from '../../actions/setupMenuActions';

class CreateMenu extends Component {
  constructor(props) {
    super(props);
  }

  componentDidMount() {
    this.props.getAMeals({limit: 20, page: 1});
  }

  handleSubmit = (values) => {
    this.props.setAMenu(values)
    this.props.closeModal();
  }

  render() {
    const {
      meals, isCreating, isFetching,
    } = this.props;

    return (
      <div className='setup-form'>
        <h2>Create a menu</h2>
        {isFetching
          ?
          <h3>Fetching meals...</h3>
          :
          <MealSetupForm
            handleSubmit={this.handleSubmit}
            meals={meals}
            action="Create"
            isSubmitting={isCreating}
            date={this.props.selectedDate}
          />}
      </div>
    );
  }
}

CreateMenu.propTypes = {
  getAMeals: PropTypes.func.isRequired,
  setAMenu: PropTypes.func.isRequired,
  meals: PropTypes.arrayOf(PropTypes.object).isRequired,
  isFetching: PropTypes.bool.isRequired,
  isCreating: PropTypes.bool.isRequired,
  closeModal: PropTypes.func.isRequired,
  selectedDate: PropTypes.string.isRequired,
};

const mapStateToProps = state => ({
  meals: state.setupMenuReducer.meals,
  isFetching: state.setupMenuReducer.isFetching,
  isCreating: state.setupMenuReducer.isCreating,
  selectedDate: state.setupMenuReducer.selectedDate,
});

export default connect(mapStateToProps, { getAMeals, setAMenu })(CreateMenu);