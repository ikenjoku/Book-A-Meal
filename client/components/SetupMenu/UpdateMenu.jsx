import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import MealSetupForm from './MealSetupForm';
import { getAMeals, updateAMenu } from '../../actions/setupMenuActions';

export class UpdateMenu extends Component {
  constructor(props) {
    super(props);
  }

  componentDidMount() {
    this.props.getAMeals({limit: 20, page: 1});
  }

  handleSubmit = (values) => {
    this.props.updateAMenu(values, this.props.menuId)
    this.props.closeModal()
  }

  render() {
    const {
      meals, isUpdating, isFetching,
    } = this.props;

    return (
      <div className='setup-form'>
        <h2>Update Menu</h2>
        {isFetching
          ?
          <p className='loading-update-meals'>Loading...</p>
          :
          <MealSetupForm
            handleSubmit={this.handleSubmit}
            meals={meals}
            action="Update"
            isSubmitting={isUpdating}
            date={this.props.selectedDate}
          />}
      </div>
    );
  }
}

UpdateMenu.propTypes = {
  updateAMenu: PropTypes.func.isRequired,
  getAMeals: PropTypes.func.isRequired,
  meals: PropTypes.arrayOf(PropTypes.object).isRequired,
  isUpdating: PropTypes.bool.isRequired,
  isFetching: PropTypes.bool.isRequired,
  menuId: PropTypes.number.isRequired,
  closeModal: PropTypes.func.isRequired,
  selectedDate: PropTypes.string.isRequired,
};

export const mapStateToProps = state => ({
  meals: state.setupMenuReducer.meals,
  isUpdating: state.setupMenuReducer.isUpdating,
  isFetching: state.setupMenuReducer.isFetching,
  menuId: state.setupMenuReducer.currentMenu.id,
  selectedDate: state.setupMenuReducer.selectedDate,
});

export default connect(mapStateToProps, { getAMeals, updateAMenu })(UpdateMenu);

