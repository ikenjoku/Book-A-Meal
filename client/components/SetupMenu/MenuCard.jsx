import React, { Component } from "react";
import moment from 'moment';
import Modal from 'react-modal';
import { connect } from "react-redux";
import { getAMenu } from "../../actions/setupMenuActions";
import CreateMenu from './CreateMenu';
import UpdateMenu from './UpdateMenu';

class MenuCard extends Component{
  constructor(props){
    super(props)
    this.state = {
      isOpen: false,
    }
  }
  handleOpenModal = () => {
    this.setState({ isOpen: true });
  }

  handleCloseModal = () => {
    this.setState({ isOpen: false });
  }

  handleMenuUpdate() {
    this.handleOpenModal();
  }

  render(){

    const modalStyle = {
      overlay: {
        backgroundColor: 'rgba(0, 0, 0, 0.25)',
      },
      content: {
        top: '50%',
        left: '50%',
        right: 'auto',
        bottom: 'auto',
        marginRight: '-50%',
        transform: 'translate(-50%, -50%)',
        minWidth: '20rem',
        width: '60%',
        overflow: 'hidden',
      },
    };

    const renderedContent = this.props.currentMenu.date ?
        <div className='menu-card'>
        <h3>Menu for { moment(this.props.currentMenu.date).format("dddd, MMMM Do YYYY") }</h3>
          <div className='menu-card-meals'>
          <ul>
            {
              this.props.currentMenu.Meals.map(meal => <li>{ meal.name }</li>)
            }
          </ul>
          </div>
          <button
            onClick={this.handleOpenModal}
          >Update Menu</button>
          <Modal
            isOpen={this.state.isOpen}
            contentLabel="Menu"
            style={modalStyle}
          >
            <button
              value="&times;"
              onClick={this.handleCloseModal}
              className="close"
            >close</button>
             <UpdateMenu closeModal={this.handleCloseModal} />
          </Modal>
        </div> :
          <div className='no-menu-card'>
            <h3>No Menu has been set for this day</h3>
            <button
              onClick={this.handleOpenModal}
            >Create Menu</button>
            <div>
              <Modal
              isOpen={this.state.isOpen}
              contentLabel="Menu"
              style={modalStyle}
            >
              <button
                value="&times;"
                onClick={this.handleCloseModal}
                className="close"
              >close</button>
                <CreateMenu closeModal={this.handleCloseModal} />
            </Modal>
            </div>
          </div> 
    return(
      <div className=''>
        {renderedContent}
      </div>
    );
  }
}

const mapStateToProps = state => {
  return {
    currentMenu: state.setupMenuReducer.currentMenu,
    isLoadingMenu: state.setupMenuReducer.isLoadingMenu,
    error: state.setupMenuReducer.error
  }
}

export default connect(mapStateToProps, { getAMenu })(MenuCard);