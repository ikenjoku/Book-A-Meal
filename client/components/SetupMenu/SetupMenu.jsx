import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import DatePicker from '../DatePicker';
import MenuCard from './MenuCard';
import { getAMenu, setSelectedDate } from '../../actions/setupMenuActions';

export class SetupMenu extends Component {
  componentDidMount() {
    const today = new Date().toISOString().substr(0, 10);
    this.props.getAMenu({ selectedDate: today });
    this.props.setSelectedDate({ selectedDate: today });
  }

  render() {
    return (
      <main id="setup-menu-container">
        <div className="cool-lg-text">
          <h2>Setup Menu</h2>
        </div>
        <DatePicker
          btnName="Find Menu"
          onSubmit={selectedDate => this.props.getAMenu(selectedDate)}
          onChange={selectedDate => this.props.setSelectedDate(selectedDate)}
        />
        <MenuCard />
      </main>
    );
  }
}

SetupMenu.propTypes = {
  getAMenu: PropTypes.func.isRequired,
  setSelectedDate: PropTypes.func.isRequired,
};

export default connect(null, { getAMenu, setSelectedDate })(SetupMenu);
