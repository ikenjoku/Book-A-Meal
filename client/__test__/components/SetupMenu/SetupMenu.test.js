import React from 'react';
import { shallow } from "enzyme";
import { SetupMenu,  } from "../../../components/SetupMenu/SetupMenu";

describe('SetupMenu Component', () => {
  let getAMenu, wrapper, setSelectedDate;
  getAMenu = jest.fn();
  setSelectedDate = jest.fn();

  test('should render the SetupMenu correctly', () => {
    wrapper = shallow(<SetupMenu getAMenu={getAMenu} setSelectedDate={setSelectedDate} />);
    expect(wrapper).toMatchSnapshot();
  });
  
  test('DatePicker should handle onSubmit', () => {
    wrapper.find('DatePicker').prop('onSubmit')('2018-08-08');
    expect(getAMenu).toHaveBeenLastCalledWith('2018-08-08');
  });

  test('DatePicker should handle onChange', () => {
    wrapper.find('DatePicker').prop('onChange')('2018-08-08');
    expect(setSelectedDate).toHaveBeenLastCalledWith('2018-08-08');
  });
});
