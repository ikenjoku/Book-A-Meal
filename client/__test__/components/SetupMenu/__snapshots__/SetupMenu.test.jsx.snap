// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SetupMenu Component should render the SetupMenu correctly 1`] = `
<main
  id="setup-menu-container"
>
  <div
    className="cool-lg-text"
  >
    <h2>
      Setup Menu
    </h2>
  </div>
  <DatePicker
    btnName="Find Menu"
    onChange={[Function]}
    onSubmit={[Function]}
  />
  <Connect(MenuCard) />
</main>
`;
